### PR TITLE
Fix check for sensitive and restricted domains

### PR DESCRIPTION
**Pull request by intgr**
This is a fix for a potential security issue. Fix submitted as a pull request per discussion with @kevgliss.

There was much wrong with this code:
* If there were no matches in the Domain table then
  LEMUR_RESTRICTED_DOMAINS was ignored.
* If there were multiple Domain matches, the LEMUR_RESTRICTED_DOMAINS
  checks were ran multiple times.
* It's unnecessary to query Domain rows if the user is an admin.
* If a domain failed the check, the error would say
  "Domain Domain(name=example.org) has been marked as sensitive ..."

### CODE REVIEW

File: lemur/common/validators.py [hunk 1]
```diff
@@ -42,20 +42,20 @@ def private_key(key):
...
-        domains = domain_service.get_by_name(domain)
-        for domain in domains:
-            # we only care about non-admins
-            if not SensitiveDomainPermission().can():
-                if domain.sensitive or any([re.match(pattern, domain.name) for pattern in restricted_domains]):
-                    raise ValidationError(
-                        'Domain {0} has been marked as sensitive, contact and administrator \
-                        to issue the certificate.'.format(domain))
+    if any(re.match(pattern, domain) for pattern in restricted_domains) or \
+            any(d.sensitive for d in domain_service.get_by_name(domain)):
```
Comment (intgr) [inline lemur/common/validators.py]: This could be rewritten to filter Domain.sensitive on database-side. But I guess the old code was written in this style to keep the generality of `get_by_name()`.

File: lemur/common/validators.py [hunk 2]
```diff
@@ -42,20 +42,20 @@ def private_key(key):
...
-    Determines if domain has been marked as sensitive.
+    Determines if domain has been marked as sensitive or matches restricted patterns.
     :param domain:
     :return:
     """
+    if SensitiveDomainPermission().can():
+        # User has permission, no need to check anything
+        return
+
     restricted_domains = current_app.config.get('LEMUR_RESTRICTED_DOMAINS', [])
```
Comment (kevgliss) [inline lemur/common/validators.py]: For us, we generally allow all domains (as validated by issuers) and only blacklist them very specifically. I think at some point we should do away with this setting altogether and rely on the `sensitive` attribute on the domain model. The domains UI in particular has been neglected so it might take a bit of work to get it going.

File: lemur/common/validators.py [hunk 3]
```diff
@@ -42,20 +42,20 @@ def private_key(key):
 
 def sensitive_domain(domain):
     """
-    Determines if domain has been marked as sensitive.
+    Determines if domain has been marked as sensitive or matches restricted patterns.
     :param domain:
     :return:
     """
+    if SensitiveDomainPermission().can():
+        # User has permission, no need to check anything
```
Comment (kevgliss) [inline lemur/common/validators.py]: We used to have something like this:

https://github.com/Netflix/lemur/blob/0.4.0/lemur/tests/conftest.py#L183

I had removed that fixture when I removed most of the permission logic from the services and into the views. This may be an edge case where this particular fixture is useful again.

File: lemur/common/validators.py (hunk 1 above)
Comment (kevgliss) [inline lemur/common/validators.py]: It depends, we can either keep the regex matching or rely entirely on the domain.sensitive field for this logic. I would prefer the latter going forward as I believe it would be more flexible.

**Comment by kevgliss** [comment]
If we choose to not only fix this functionality, but improve it. I think adding a event log similar to: https://github.com/Netflix/lemur/blob/master/lemur/certificates/views.py#L449
when a request is rejected might make sense for an auditing perspective.

**Comment by kevgliss** [comment]
@intgr Any update on this PR?

**Comment by intgr** [comment]
@kevgliss Sorry, I've been busy with other things. Hopefully next week I can return to working on Lemur.

**Comment by kevgliss** [comment]
Closing this now for now. Feel free to re-open.
